packages/checkout-storefront/src/sections/PaymentSection/AdyenDropIn/useAdyenDropin.ts

```diff
@@ -0,0 +1,321 @@
+import { usePaymentProcessingScreen } from "@/checkout-storefront/sections/PaymentSection/PaymentProcessingScreen";
+
+export interface AdyenDropinProps {
+  config: ParsedAdyenGateway;
+}
+
+export const useAdyenDropin = (props: AdyenDropinProps) => {
+  const { config } = props;
+  const { id } = config;
+
+  const {
+    checkout: { id: checkoutId, totalPrice },
+  } = useCheckout();
+  const { authenticated } = useUser();
+  const { errorMessages } = useErrorMessages(adyenErrorMessages);
+  const { errorMessages: commonErrorMessages } = useErrorMessages(apiErrorMessages);
+  const { validateAllForms } = useCheckoutValidationActions();
+  const { validationState } = useCheckoutValidationState();
+  const { updateState, loadingCheckout, ...rest } = useCheckoutUpdateState();
+  const { showCustomErrors } = useAlerts();
+  const { submitInProgress } = useCheckoutUpdateState();
+  const { setSubmitInProgress } = useCheckoutUpdateStateActions();
+  const { setIsProcessingPayment } = usePaymentProcessingScreen();
+
+  const [currentTransactionId, setCurrentTransactionId] = useState<ParamBasicValue>(
+    getQueryParams().transaction
+  );
+  const [, transactionInitialize] = useTransactionInitializeMutation();
+  const [, transactionProccess] = useTransactionProcessMutation();
+  const { onCheckoutComplete } = useCheckoutComplete();
+
+  const [adyenCheckoutSubmitParams, setAdyenCheckoutSubmitParams] = useState<{
+    state: AdyenCheckoutInstanceState;
+    component: DropinElement;
+  } | null>(null);
+
+  const anyRequestsInProgress = areAnyRequestsInProgress({ updateState, loadingCheckout, ...rest });
+
+  const finishedApiChangesWithNoError = hasFinishedApiChangesWithNoError({
+    updateState,
+    loadingCheckout,
+    ...rest,
+  });
+
+  const handlePaymentResult = useCallback(
+    ({
+      paymentResponse,
+      transaction,
+    }: {
+      paymentResponse: AdyenPaymentResponse;
+      transaction: MightNotExist<{ id: string }>;
+    }) => {
+      const { action, resultCode } = paymentResponse;
+
+      if (transaction) {
+        setCurrentTransactionId(transaction.id);
+        replaceUrl({ query: { transaction: transaction.id } });
+      }
+
+      if (action) {
+        adyenCheckoutSubmitParams?.component.handleAction(action);
+      }
+
+      switch (resultCode) {
+        case "Authorised":
+          adyenCheckoutSubmitParams?.component.setStatus("success");
+          void onCheckoutComplete();
+          return;
+        case "Error":
+          adyenCheckoutSubmitParams?.component.setStatus("error");
+          showCustomErrors([{ message: "There was an error processing your payment." }]);
+          return;
+        case "Refused":
+          setCurrentTransactionId(undefined);
+
+          adyenCheckoutSubmitParams?.component.setStatus("ready");
+
+          const messageKey = camelCase(paymentResponse.refusalReason);
+
+          showCustomErrors([{ message: errorMessages[messageKey as keyof typeof errorMessages] }]);
```

Comment:
What if the error message is not included in that object? Do we have some default?

Reply:
Good catch, will add a condition for that